ref/haskell/src/Codec/Binary/Bech32.hs

```diff
@@ -0,0 +1,128 @@
+module Codec.Binary.Bech32
+  ( bech32Encode
+  , bech32Decode
+  , segwitEncode
+  , segwitDecode
+  , Word5(UnsafeWord5)
+  ) where
+
+import Control.Monad (guard)
+import Data.Bits (Bits, unsafeShiftL, unsafeShiftR, (.&.), xor, testBit)
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import Data.Char (toLower, toUpper)
+import Data.Digits
+import Data.Foldable (foldl')
+import qualified Data.HashMap.Strict as Map
+import Data.Word (Word8, Word32)
+import Data.Hashable (Hashable(..))
+
+type HRP = BS.ByteString
+type Data = [Word8]
+
+(.>>.), (.<<.) :: Bits a => a -> Int -> a
+(.>>.) = unsafeShiftR
+(.<<.) = unsafeShiftL
+
+newtype Word5 = UnsafeWord5 Word8
+              deriving (Eq)
+
+instance Hashable Word5 where
+    hashWithSalt salt (UnsafeWord5 x) = hashWithSalt salt x
+
+word5 :: Word8 -> Word5
+word5 x = UnsafeWord5 (x .&. 31)
+
+fromWord5 :: Num a => Word5 -> a
+fromWord5 (UnsafeWord5 x) = fromIntegral x
+
+charset :: String
+charset = "qpzry9x8gf2tvdw0s3jn54khce6mua7l"
```

Comment:
```
charset :: Array Word5 Char
charset = listArray (0,31) "qpzry9x8gf2tvdw0s3jn54khce6mua7l"

charsetMap :: Char -> Maybe Word5
charsetMap c | upperC `inRange` bounds invArray = invArray ! upperC
             | otherwise = Nothing
 where
  upperC = toUpper c
  invArr :: Array Char (Maybe Word5)
  invArr = accumArray ('0','Z') (const Just) Nothing (map swap (assocs charset))
   where
    swap (a,b) = (b,a)

charsetMapInv :: Word5 -> Char
charsetMapInv i = charset ! i